### PR TITLE
Add FRF noise covariance calculation library

### PR DESCRIPTION
**Pull request by mwilensky768**
This branch adds some noise propagation functions for the fringe rate filter + coherent averaging steps. 

I've added the basic functions that are demonstrated in the single baseline power spectrum notebook, though they've been modified here to be slightly more general since the notebook versions used notebook-specific info. This PR and [the corresponding one on that repo](https://github.com/HERA-Team/hera_notebook_templates/pull/54) should probably coordinate.

I still need to add unit tests and think I should add a wrapper that goes from ee + nn autos/crosses all the way to a pI time-time cov. per freq so that there is a clearly implemented "pipeline" somewhere, though maybe that should just live in another repo as part of actual pipeline code.

### CODE REVIEW

File: hera_cal/frf.py [hunk 1]
```diff
@@ -1998,3 +1999,245 @@ def time_average_argparser():
...
+            Whether to coherently average on the t_avg cadence or not.
+        wgt_tavg_by_nsample (bool):
+            Whether to weight the time average by nsamples. 
+            Otherwise do uniform weighting.
+        nsamples (array):
+            Array proportional to how many nights contributed to each sample.
+        rephase (bool):
+            Whether to rephase before averaging.
+        bl_vec (array):
+            Baseline vector in ENU
```
Comment (jsdillon) [inline hera_cal/frf.py:2038]: in meters, right?

File: hera_cal/frf.py [hunk 2]
```diff
@@ -1998,3 +1999,245 @@ def time_average_argparser():
...
+
+    Returns:
+        var: array
+            Noise variance (Ntimes, Nfreqs)
+    """
+    var = predict_noise_variance_from_autos(cross_antpairpol, data,
+                                            nsamples=nsamples, auto_ant=auto_ant)
+    var = var[:, freq_slice]
+
+    var_isinf = np.isinf(var)
```
Comment (jsdillon) [inline hera_cal/frf.py:2163]: you might want to is `~np.isfinite()` unless for some reason you don't want to treat nans as infinite variance (or you're very confident you won't have any nans)

File: hera_cal/frf.py [hunk 3]
```diff
@@ -1998,3 +1999,245 @@ def time_average_argparser():
...
+    Ntimes = cov.shape[1]
+    diags = cov[:, np.arange(Ntimes), np.arange(Ntimes)]
+    corr = cov / np.sqrt(diags[:, None] * diags[:, :, None])
+    
+    return corr
+
+def get_correction_factor_from_cov(cov, tslc=None):
+    """
+    Get a correction factor for PS noise variance prediction in the absence
+    of propagating the noise covariances all the way to delay space.
```
Comment (jsdillon) [inline hera_cal/frf.py:2219]: Do you want to say a bit more about how this factor is defined?

File: hera_cal/tests/test_frf.py [hunk 4]
```diff
@@ -1066,3 +1067,241 @@ def test_tophat_linear_argparser(self):
...
+
+
+    
+
+
+
+
+    
+
+
```
Comment (jsdillon) [inline hera_cal/tests/test_frf.py:1307]: don't think you meant to add all this whitespace

File: hera_cal/noise.py [hunk 5]
```diff
@@ -92,7 +98,11 @@ def predict_noise_variance_from_autos(bl, data, dt=None, df=None, nsamples=None)
         df = np.median(np.ediff1d(data.freqs))
 
     ap1, ap2 = split_pol(bl[2])
-    auto_bl1, auto_bl2 = (bl[0], bl[0], join_pol(ap1, ap1)), (bl[1], bl[1], join_pol(ap2, ap2))
+    if auto_ant is None:
+        auto_bl1, auto_bl2 = (bl[0], bl[0], join_pol(ap1, ap1)), (bl[1], bl[1], join_pol(ap2, ap2))
+    else:
+        auto_bl1, auto_bl2 = (auto_ant, auto_ant, join_pol(ap1, ap1)), (auto_ant, auto_ant, join_pol(ap2, ap2))
+
```
Comment (jsdillon) [inline hera_cal/noise.py:105]: Can you add a little unit test for this new functionality?

**Review by jsdillon — CHANGES_REQUESTED** [review]
Thanks @mwilensky768. This all looks basically fine, just a couple of minor suggestions

File: hera_cal/frf.py [hunk 6]
```diff
@@ -1998,3 +1999,245 @@ def time_average_argparser():
...
+            (Interleaved) Julian Dates on hd, converted to seconds.
+        filter_cent_use (float):
+            Filter center for main lobe filter, in Hz.
+        filter_half_wid_use (float):
+            Filter half width for main lobe filter, in Hz.
+        freqs (array):
+            Frequencies in the data (in Hz).
+        t_avg (float):
+            Desired coherent averaging length, in seconds.
+        cutoff (float):
```
[the remark this answers is not in the document]
Comment (mwilensky768) [reply hera_cal/frf.py]: Done

File: hera_cal/frf.py (hunk 1 above)
Comment (mwilensky768) [reply hera_cal/frf.py]: Yep, added this to the docstring

File: hera_cal/frf.py (hunk 2 above)
Comment (mwilensky768) [reply hera_cal/frf.py]: Thanks, that's a much safer suggestion. I've implemented it.

File: hera_cal/frf.py (hunk 3 above)
Comment (mwilensky768) [reply hera_cal/frf.py]: I've now pointed the reader to the HERA memo and explained the memo in one sentence.

File: hera_cal/tests/test_frf.py (hunk 4 above)
Comment (mwilensky768) [reply hera_cal/tests/test_frf.py]: Correct -- rampant pressing of the enter key. I've removed it.

File: hera_cal/noise.py (hunk 5 above)
Comment (mwilensky768) [reply hera_cal/noise.py:105]: This new functionality is checked in `test_prep_var_for_frop` -- do you have a different way you'd like to check it? The only difference in cutting out the `prep_var_for_frop` wrapper is whether nonfinite values get resassigned.

Comment (mwilensky768) [reply hera_cal/noise.py:105]: I know it's sort of funky design to test something in `noise` in `test_frf`

Comment (jsdillon) [reply hera_cal/noise.py:105]: meh, I think that's good enough

**Review by jsdillon — APPROVED** [review]
Assuming tests pass... Thanks @mwilensky768 !

**Pull request merged**